app/crud/serviceMetric.py

```diff
@@ -1,19 +1,142 @@
-
-from sqlalchemy.orm import Session
-
-from ..models import ServiceMetric
-from ..schemas import ServiceMetricCreate, ServiceMetricUpdate
+from sqlalchemy.orm import Session 
+from sqlalchemy import func, and_
+from ..models import ServiceMetric, scoreCardMetrics
+from ..schemas import ServiceMetricCreate, ServiceMetricUpdate , ServiceMetricCL
 from .base import CRUDBase
-
+from . import CRUDScoreCardMetric, CRUDMetric
+from app.utils.utility_functions import main_calculate_error, calculate_score
+from typing import Optional
+from datetime import datetime
 
 class CRUDServiceMetric(CRUDBase[ServiceMetric, ServiceMetricCreate, ServiceMetricUpdate]):
     def __init__(self, db_session: Session):
         super(CRUDServiceMetric, self).__init__(ServiceMetric, db_session)
+        self.scorecardMetrics = CRUDScoreCardMetric(db_session)
+        self.Metric = CRUDMetric(db_session)
 
     def getByScorecardId(self, scorecardId: int) -> list[ServiceMetric]:
         return self.db_session.query(ServiceMetric).filter(ServiceMetric.scorecardId == scorecardId).all()
 
     def getByServiceId(self, serviceId: int) -> list[ServiceMetric]:
         return self.db_session.query(ServiceMetric).filter(ServiceMetric.serviceId == serviceId).all()
 
+    def get_metrics_by_scorecard_and_service(self, scorecard_id: int, service_id: int) -> list[ServiceMetric]:
+        metrics = self.db_session.query(ServiceMetric).filter(
+            ServiceMetric.serviceId == service_id,
+            scoreCardMetrics.scoreCardId == scorecard_id
+        ).all()
+        return metrics
+    
+    # fun for get metric list with range time
+    def get_metric_values_by_service(self, service_id: int,from_date: Optional[datetime], to_date: Optional[datetime]) -> list[ServiceMetricCL]:
+        metrics = (
+            self.db_session.query(ServiceMetric.metricId,ServiceMetric.value,ServiceMetric.timestamp)
+        )
+        query = self.db_session.query(ServiceMetric.metricId, ServiceMetric.value, ServiceMetric.timestamp)  # Query initialized here
+        if from_date and to_date:
+          query = query.filter(ServiceMetric.timestamp >= from_date, ServiceMetric.timestamp <= to_date)
+        elif from_date:
+          query = query.filter(ServiceMetric.timestamp >= from_date)
+        elif to_date:
+          query = query.filter(ServiceMetric.timestamp <= to_date)
+        
+        query = query.filter(ServiceMetric.serviceId == service_id).order_by(ServiceMetric.timestamp.desc())
```

Comment:
suggesstion: move the filter by serviceId to the initial query at line 35. It's applied regardless whether a filter by timestamp exists or not.
but keep the order_by in its current place.
such that 
`query = self.db_session.query(ServiceMetric.metricId, ServiceMetric.value, ServiceMetric.timestamp)` becomes `query = self.db_session.query(ServiceMetric.metricId, ServiceMetric.value, ServiceMetric.timestamp).filter(ServiceMetric.serviceId == service_id)`
and ` query = query.filter(ServiceMetric.serviceId == service_id).order_by(ServiceMetric.timestamp.desc())` becomes ` query = query.order_by(ServiceMetric.timestamp.desc())`

Reply:
done